src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs

```diff
@@ -36,19 +40,24 @@ private BoundExpression VisitCompoundAssignmentOperator(BoundCompoundAssignmentO
             bool isPossibleEventHandlerOperation = node.Left.Kind == BoundKind.DynamicMemberAccess &&
                 (binaryOperator == BinaryOperatorKind.Addition || binaryOperator == BinaryOperatorKind.Subtraction);
 
+            var lhsRead = MakeRValue(transformedLHS);
+
             // save RHS to a temp, we need to use it twice:
+            // if RHS needs to be saved to a temp, we save the LHS read to a temp first to ensure order
+            // of operations is preserved
             if (isPossibleEventHandlerOperation && CanChangeValueBetweenReads(loweredRight))
             {
```

Comment:
This might fix the problem for += and -=, but not for &=, |=, etc.

Reply:
Only += and -= are affected by this. I'll add some tests for & and |, but the bug is specific to operators that can possibly be event handlers.

---
In reply to: [199892357](https://github.com/dotnet/roslyn/pull/28247#discussion_r199892357) [](ancestors = 199892357)